WIP: Add start and end dates for notices

cameron1729 opened the pull request.

dmitriim commented on db/upgrade.php:
```diff
@@ -45,5 +45,25 @@ function xmldb_local_sitenotice_upgrade($oldversion) {
+        if (!$dbman->field_exists('local_sitenotice', 'timeend')) {
+
+            $table = new xmldb_table('local_sitenotice');
+            $field = new xmldb_field('timestart', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNULL, null, '0');
```
Have you tested that upgrade? :)

dmitriim replied:
Have you tested that upgrade? :)

dmitriim commented:
Would be awesome if we can have a notice that is active from specific date and then never expires. Of course this could be achieved by setting up end date far in the future.  Just thinking out loud here.

dmitriim commented:
Also a quick note from UI testing. Do we really need Is perpetual column? I think we can save space and display only time start and time end. If they are not set, means its perpetual. Thoughts?

![image](https://user-images.githubusercontent.com/4626431/173701950-c2cc9d52-b167-409d-8d5e-abc069524729.png)

cameron1729 commented:
Thanks Dima. I've made a few changes (for some reason I can't reply directly to your comments):

1. Got rid of the "Is perpetual" column - you're right, it adds no value to the user
2. Added docblocks where needed
3. Updated unit tests
4. Fixed that upgrade step
5. Having a notice with a start date but no end date is a cool idea (and the other way around too), I'm wondering how we can implement this in the UI? Moodle forms make it a bit challenging. I think it might be a bit out of scope for this issue, but if I have any thoughts about how it could work I'll implement them, if not, I'll create an issue for it

Can you have another look?

dmitriim commented:
I pulled the latest changes and still can see "Is perpetual" column. Did you forget to push an extra commit?

dmitriim commented:
When I'm trying to save an exiting notice, I'm getting following error

![image](https://user-images.githubusercontent.com/4626431/173961513-ae58158e-6c40-4235-b240-0d147bfdc9b1.png)

![image](https://user-images.githubusercontent.com/4626431/173961449-a7305f5c-a358-49ee-ac43-98c50bb095a0.png)

dmitriim commented on tests/sitenotice_test.php:
```diff
@@ -78,6 +78,30 @@ private function create_cohort_notice2() {
+    /**
+     * Create sample notice (with a start and end date).
+     */
+    private function create_notice_with_start_and_end2() {
```
I don't think having methods like blabla1 and blabla2 is the best practice. Can you please rewrite to have one generic method where you can pass all required data from your test?

I see that the code already does that. If you can fix existing code that would be a bonus.

cameron1729 commented:
Thanks Dima, I've reworked the unit tests to use data providers. It's not perfect but it's better than it was.

I think the problem about that unexpected property is gone. I tried the following:

1. Make a notice before the patch
2. Upgrade
3. Edit the notice, adding a start and end date

And I didn't see any error.
